webrtc/RTCPeerConnection-helper.js

```diff
@@ -775,6 +775,33 @@ function findTransceiverForSender(pc, sender) {
   return null;
 }
 
+// Uses WebAudio to detect audio (or lack of audio).
+function waitForAudio(stream, silence = false) {
+  const ac = new AudioContext();
+  const analyser = ac.createAnalyser();
+  analyser.fftSize = 2048;
+  analyser.smoothingTimeConstant = 0;
+  const fftBins = new Uint8Array(analyser.frequencyBinCount);
+  const sourceNode = ac.createMediaStreamSource(new MediaStream([stream.getAudioTracks()[0]]));
+  sourceNode.connect(analyser);
+  return new Promise(resolve => {
+    setTimeout(function poll() {
+      analyser.getByteFrequencyData(fftBins);
+	  const sum = fftBins.reduce((a, b) => a + b);
+	  if ((silence && sum === 0) || (!silence && sum > 0)) {
+         ac.close().then(resolve);
+      } else {
+	    requestAnimationFrame(poll);
+      }
+    }, 500);
+  });
+}
+
```

Comment:
See also https://stackoverflow.com/a/46781986, and https://stackoverflow.com/a/40691112 using `AudioContext` clock. Note, detecting for silence depends heavily on context. One case where attempting to detect silence can be evaluated as a false positive is SSML input to TTS that is then used for STT can include `<break>` and `<break time="2500" strength="medium">` elements, where if, for example if there is code the executes `stop()` on a `MediaStreamTrack` when silence is detected the media stream could still be live and only outputting silence based on markup commanding the application to do so, for arbitrary time spans. That is one issue with attempting to implement TTS to `MediaStreamTrack`, solvable to some extent by trying to seek to EOF, difficult for a live TTS output: when to stop the stream when silence is intended? 

Does `requestAnimationFrame()` require focus, or have the potential to change behaviour when focus is not on the tab?